Revamp config management

llucax opened the pull request:
This pull request adds a new `ConfigManager` class that takes care of the *client-side* of the `ConfigManagingActor`, providing the common tasks users of the config system need to do.

The class owns the config channel and provides a convenient method to get new receivers that validate the configuration using a `dataclass` as a `marshmallow` schema, and allow filtering to receive only configuration for a specific key. The receiver also reports errors when the configuration can't be properly loaded, and avoid spurious updates when the configuration didn't change compared to the previous one.

The manager also takes care of instantiating both the `ConfigManagingActor` and the `LoggingConfigUpdatingActor` and provides a utility function to wait for the first configuration to be received.

llucax committed:
Add a `ConfigManager` class

This class instantiates and starts the `ConfigManagingActor` and creates
the channel needed to communicate with it. It offers a convenience
method to get receivers to receive configuration updates.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add option to only send changed configurations

When this option is enabled, configurations will be sent to the receiver
only if they are different from the previously received configuration.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Bump the minimum required channels version to v1.4.0

This is to be able to use type guards in with `Receiver.filter` to
narrow the received type and the new `WithPrevious` class.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Subscribe to specific configuration keys

This forces a better configuration structure, by always having separate
sections for separate entities, and avoids spurious updates and noise
in the received configuration for other parties.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a config user guide and improve the documentation

Add module-level documentation to the `frequenz.sdk.config` module and
include it in the generated documentation's user guide.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add support to filter a sub-key

Configuration can be nested, and actors could have sub-actors that need
their own configuration, so we need to support filtering the
configuration by a sequence of keys.

For example, if the configuration is `{"key": {"subkey": "value"}}`, and
the key is `["key", "subkey"]`, then the receiver will get only
`"value"`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a base config schema that provides quantities support

This schema is provided to use as a default, and might be extended in
the future to support more commonly used fields that are not provided
by marshmallow by default.

To use the quantity schema we need to bump the `frequenz-quantities`
dependency and add the optional `marshmallow` dependency.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make the `LoggingConfigUpdatingActor` use the `ConfigManager`

The `LoggingConfigUpdatingActor` can now also receive `None` as
configuration (for example if the configuration is removed), in which
case it will go back to the default configuration.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Allow configuring logging via `ConfigManager`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve handling of unknown fields in the configuration

If `unkown` is not specified in the `marshmallow_load_kwargs`, it will
default to `marshmallow.EXCLUDE` instead of `marshmallow.RAISE`, as this
is what most users will need when loading to a dataclass.

But when `marshmallow.EXCLUDE` is used, a warning will be logged if
there are fields in the configuration that are being excluded, so the
user can still be aware of it and catch typos in the configuration file.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Validate and load configurations using a config dataclass

This commit adds support for validating configurations with a schema.
This is useful to ensure the configuration is correct and to receive it
as an instance of a dataclass.

The `new_receiver` method now requires a `config_class` argument that is
used to validate the configuration. If the configuration doesn't pass
the validation, it will be ignored and an error will be logged.

The schema is expected to be a dataclass, which is used to create a
marshmallow schema to validate the configuration. To customize the
schema, you can use `marshmallow_dataclass` to specify extra metadata.

Validation errors are logged and the `ValidationError` instance is sent
to the receiver.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make `ConfigManagingActor` logging less redundant

We were using `str(self)` as prefix, which prints the class name, but
the logger name already has enough information to figure out this log
comes from the config managing actor, so we now use only the unique ID
(name) of the actor in the logs.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a new utility function to wait for the first config

This function has a default timeout, so waiting for the first
configuration can be automatically aborted when it takes too long.

The method will also reports some progress, logging errors if there were
issues receiving or loading the configuration, or if the configuration
is missing, setting a logging baseline for all actors.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use `wait_for_first` in the `LoggingConfigUpdatingActor`

We now wait for the first configuration using the new `wait_for_first()`
utility function.

We also make the logging more explicity about when the old configuration
is being kept and why.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
This is a draft because I need to complete/improve documentation and add tests, but the code should be pretty solid, and the existing tests pass (the CI is red because I left a couple of `TODO`s that make both `pylint` and `pytest` fail).

* Based on #1133
* Replaces #1104

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_manager.py:
```diff
@@ -133,26 +164,72 @@ def new_receiver(
+            except InvalidValueForKeyError as error:
+                return error
+
+        key_receiver = receiver.map(_get_key_or_error)
```
I think it would be more clear (and it would simplify other methods) if you finish processing `key` here completely ( so it doesn't influence other methods )
You can do that by adding one extra filter: " IsNotError()"

```
def isNotErro(config):
   if isinstance(new_value, InvalidValueForKeyError):
      _logger.error
      return False
   return True 
```

 `key_receiver = receiver.map(_get_key_or_error).filter(isNotError)`

llucax replied:
This was changed in subsequent commits, does this comment still applies? Maybe you can make it in the end full diff instead so the context makes more sense?

ela-kotulska-frequenz replied:
I meant something like this:
```
        receiver = self.config_channel.new_receiver(name=f"{self}:{key}", limit=1)
        receiver = receiver.map(_get_key_or_error).filter(is_not_exception)
        receiver = receiver.map(_load_config_or_error).filter(is_not_exception)
        if skip_unchanged:
            receiver = receiver.filter(WithPrevious[DataclassT | None](
                lambda old, new: _not_equal_with_logging(
                    key=key, old_value=old, new_value=new
                )
            ))
```
Then all following method will simplify , you don't have to pass Exception as argument and check for exception. And typing will simplify, too because filter(is_not_exception) will stop as soon any exception appears.

But I just found outthat the receiver returns InvalidValueForKeyError exception, why is it necessary?
`Receiver[Mapping[str, Any] | InvalidValueForKeyError | None]:`

llucax replied:
The original PR did what you are suggesting (completely hide config loading errors from the downstream users), but I thought doing that might not be the best approach, maybe downstream actors would want to do something with this information (an error and what kind of error while trying to get the new configuration), like falling back to some safe default config or something, the current approach of keeping the old configuration is only one possible way to handle errors, but errors are still there and people might want to handle them differently.

If actors want to keep the old config if there are errors retrieving the new config, they can do it explicitly, it is just 2 extra lines of code. I also think it is good to see at the actor level more explicitly that we are handling possible errors.

I would at some point even make the `ConfigManagingActor` also send exceptions via the channel, like errors reading the file or parsing the TOML, that right now are completely being hidden to the users.

This was the main motivation for this.

ela-kotulska-frequenz replied:
Ok, you are right, it is better to return it.  Thanks for explanation!

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_manager.py:
```diff
@@ -133,26 +164,72 @@ def new_receiver(
+    Raises:
+        InvalidValueForKeyError: If the value under the key is not a mapping.
+    """
+    match config.get(key):
```
It can raise KeyError, is it ok?
If yes, could you please add comment why as it is not obvious to me :) 
If no, Maybe raising InvalidValueForKeyError for `KeyError`?

llucax replied:
No, `.get()` returns `None` if not found, so it should be fine.

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_manager.py:
```diff
@@ -175,44 +207,49 @@ def new_receiver(
-                return error
-
-        key_receiver = receiver.map(_get_key_or_error)
+        receiver = self.config_channel.new_receiver(name=f"{self}:{key}", limit=1).map(
+            lambda config: _load_config_with_logging_and_errors(
+                config,
+                config_class,
+                key=key,
+                base_schema=base_schema,
+                marshmallow_load_kwargs=marshmallow_load_kwargs,
+            )
+        )
```
The same, please add extra filter `IsNotException`, then this part will be completely seprate from others and you won't have to change  `_not_equal_with_logging` :)

llucax replied:
I don't fully understand your suggestion here, can you maybe post a small diff with the suggested changes?

Just for context, I tried not to apply many filters on purpose, to avoid the extra overhead, but if it really makes the code cleaner, I guess we can afford it anyways, config updates should be pretty seldom.

ela-kotulska-frequenz replied:
here :) 
https://github.com/frequenz-floss/frequenz-sdk-python/pull/1134#discussion_r1891424080

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_managing_actor.py:
```diff
@@ -185,8 +187,8 @@ async def _run(self) -> None:
             async for event in file_watcher:
                 if not event.path.exists():
                     _logger.error(
-                        "%s: Received event %s, but the watched path %s doesn't exist.",
-                        self,
+                        "[%s] Received event %s, but the watched path %s doesn't exist.",
+                        self.name,
```
is  `self.name` really needed in these logs? 
We don't use it like this anywhere and Logger includes name of the file in logs.

llucax replied:
Yeah, I guess you are right. The idea was that there could be many config managing actors, and without the name we don't know which one is actually emitting the logs, but since this will be probably always a singleton in one app, maybe is not really useful. Sometimes it is a delicate balance between the use we give this now and the use other users might give it in the future. If someone, for some reason, needs to instantiate more than one config managing actor, having the name is definitely useful.

BTW, the `name` exists solely for this (logging), but for singleton actors is not helping as there is no ambiguity.

llucax replied:
I'm undecided about this, so unless there is a strong push-back I prefer to leave it as it is now, we can change it in a future PR.

llucax commented:
If there are no major complains about this approach, I will move forward to clean it up so we can merge it soon. I will start doing that, so if anyone wants to review this, please do it sooner than later @frequenz-floss/python-sdk-team .

llucax committed:
Add tests for the `ConfigManager`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
> * Replaces #1104 

Maybe too late for this, but these are the main changes compared to the old PR (I had them written down and forgot to put them in the PR when I created it):

- Make `unknown=EXCLUDE` the default in `ConfigManager`
- Make config manager a background service
- Remove auto start (requires the constructor to be run in an async context, not useful)
- Remove `skip_none`
- Make passing of marshmallow arguments explicit
- Make `key` mandatory
- Make wrong value for key an error we send and send all errors
- Catch unexpected exceptions
- Add a new `wait_for_first` utility function

llucax committed:
Add missing cross-referencing for frequenz-quantities

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Allow passing a single configuration file

Both the `ConfigManager` and the `ConfigManagingActor` can now take a
single configuration file, both as a `str` or a `Path`.

This also fixes a bug where `str` was still accepted, but taken as a
`Sequence[str]`. This is a known issue with Python typing:
https://github.com/python/mypy/issues/11001

And while we are it, we now also explicitly reject empty sequences,
raising a `ValueError` when one is passed.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, this is ready for review. I completed the release notes, added tests, added comprehensive documentation, including recommended practices when writing actors and apps (please review!) and a user guide, and fixed a bug in the actor (and manager) when passing a string as a config file (now both accept also a single file).

I prepared the release notes to release after this is merged.

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_manager.py:
```diff
@@ -0,0 +1,118 @@
+            name: A name to use when creating actors. If `None`, `str(id(self))` will
+                be used. This is used mostly for debugging purposes.
+            polling_interval: The interval to poll for changes. Only relevant if
+                polling is enabled.
```
Not very related, but why `force_polling` and `polling_interval` should be configurable here?
1. It depends from the system whether we need `force_polling` or not.
2. `polling_interval` is ok I guess,  but description `Only relevant if polling is enabled.`  seems not correct.  Even if `force_pooling` set to False, file polling will still be used as a fallback when the underlying file system does not support event-based notifications.

llucax replied:
In general, since we are building a SDK that anyone could use on any stack, IMHO it makes sense to allow full customization. We only use `force_polling` as a safer choice as `inotify` doesn't work well on all filesystems and mount options, but if you know what you are doing, `inotify` could be more efficient and someone might want to use it instead of force-polling.

I think the *Only relevant if polling is enabled* is correct, it doesn't say *Only relevant if `force_polling` is enabled*, it talks about *polling* in general, and as you say, if there is no better way to monitor for file changes, *polling* will be enabled as a fallback even if `force_polling` is not enabled.

I would leave it as it is, it is very niche and I think it is not worth spending a lot of time writing a ton of documentation about it, if someone needs to tune this, they will probably need to go read the docs up to `watchfiles`, as it is not trivial to figure out how this works and all the subtle behavior changes that can come up from different backends and options.

ela-kotulska-frequenz commented on src/frequenz/sdk/config/_manager.py:
```diff
@@ -313,6 +315,99 @@ def new_receiver(  # pylint: disable=too-many-arguments
+    except asyncio.TimeoutError:
+        _logger.error("%s: No configuration received in time.", receiver_name)
+        raise
+    raise ReceiverStoppedError(receiver)
```
About `raise ReceiverStoppedError(receiver)`
If we plan to start and stop actors dynamically then I guess it will be be normal for receiver to stop. For example if sender stopped then receiver will probably stop, too.
And user can't really do anything if receiver stopped.
Maybe it would also be ok to just log warning, and return?

llucax replied:
Mmm, I'm not sure how would that work. If we return, it means the receiver is ready, but it is not, it is stopped. Normally this exception will bubble up and will end up restarting actors (or exiting the application if it was at the app level), which I guess is what we want? At least if the receiver is re-created when an actor is restarted.

I think there is a bigger topic here, how do we want to handle actor restarts in terms of completely resetting the state (like re-creating receivers) or not. Because of things like this, I think we should probably re-create the state, otherwise a stopped receiver will never be fixed unless each and every actor handle the exception and re-creates the receiver manually.

But this is a much wider topic and I think it is out of the scope of this PR.

ela-kotulska-frequenz reviewed (commented):
LGTM :)
